[PATCH] extmsg: fix Message-ID search for non-xap_helper users - Eric Wong
user/dev discussion of public-inbox itself
help / color / mirror / code / Atom feed
From: Eric Wong <e@80x24.org>
To: Kyle Meyer <kyle@kyleam.com>
Cc: meta@public-inbox.org
Subject: [PATCH] extmsg: fix Message-ID search for non-xap_helper users
Date: Tue, 21 Oct 2025 03:03:46 +0000	[thread overview]
Message-ID: <20251021030346.M346444@dcvr> (raw)
In-Reply-To: <20251020193141.M515023@dcvr>
> > Kyle Meyer <kyle@kyleam.com> wrote:
> > > Here's a trimmed down reproducer.
Thanks again, I just used your test as-is to write the test
case:
--------8<--------
Subject: [PATCH] extmsg: fix Message-ID search for non-xap_helper users
commit c3d0295bd was intended to avoid head-of-line blocking for
--xapian-helpers (-X) deployments.  Unfortunately, that broke
searches for deployments not using --xapian-helpers by causing
the ExtMsg to get queued redundantly into the event loop.  So
perform the `requeue' procedure exactly once and simplify
ExtMsg->event_step by eliminating the $sync parameter.
Fixes: c3d0295bd (www: extmsg: async partial Message-ID search, 2025-05-12)
Reported-by: Kyle Meyer <kyle@kyleam.com>
Link: https://public-inbox.org/meta/877bwva3s4.fsf@kyleam.com/
---
lib/PublicInbox/ExtMsg.pm        |  8 ++--
t/extmsg_event_step_regression.t | 70 ++++++++++++++++++++++++++++++++
2 files changed, 73 insertions(+), 5 deletions(-)
create mode 100644 t/extmsg_event_step_regression.t
diff --git a/lib/PublicInbox/ExtMsg.pm b/lib/PublicInbox/ExtMsg.pm
index 282db1d26..3e9b33a37 100644
--- a/lib/PublicInbox/ExtMsg.pm
+++ b/lib/PublicInbox/ExtMsg.pm
@@ -137,7 +137,7 @@ sub partial_enter ($) {
my ($ctx) = @_;
bless $ctx, __PACKAGE__; # for ExtMsg->event_step
return $ctx->event_step if $ctx->{env}->{'pi-httpd.app'};
-	$ctx->event_step(1) while $ctx->{-wcb};
+	$ctx->event_step while $ctx->{-wcb}; # generic PSGI
}
sub partial_prepare ($@) {
@@ -193,12 +193,10 @@ sub ext_msg {
# called via PublicInbox::DS::event_loop
sub event_step {
-	my ($ctx, $sync) = @_;
+	my ($ctx) = @_;
# can't find a partial match in current inbox, try the others:
my $ibx = shift @{$ctx->{again}} or return finalize_partial($ctx);
-	unless (partial_ibx_start($ctx, $ibx)) {
-		PublicInbox::DS::requeue($ctx) unless $sync;
-	}
+	partial_ibx_start $ctx, $ibx;
}
sub finalize_exact {
diff --git a/t/extmsg_event_step_regression.t b/t/extmsg_event_step_regression.t
new file mode 100644
index 000000000..941d11ae8
--- /dev/null
+++ b/t/extmsg_event_step_regression.t
@@ -0,0 +1,70 @@
+#!perl -w
+# Copyright (C) all contributors <meta@public-inbox.org>
+# License: AGPL-3.0+ <https://www.gnu.org/licenses/agpl-3.0.txt>
+use v5.12;
+use PublicInbox::TestCommon;
+use PublicInbox::Eml;
+use PublicInbox::IO qw(write_file);
+require_mods qw(-httpd v2 psgi);
+my $tmpdir = tmpdir;
+my $one = PublicInbox::Eml->new(<<'EOM');
+From: X <x@example.com>
+Subject: subj
+Date: Sat, 16 Jun 2007 17:33:53 +0100
+Message-ID: <681fb71bcf6@example.com>
+To: list@example.com
+Mime-Version: 1.0
+Content-Type: text/plain
+
+abc
+EOM
+my $two = PublicInbox::Eml->new(<<'EOM');
+From: X <x@example.com>
+Subject: Re: subj
+Date: Sat, 30 Jun 2007 18:56:36 +0100
+Message-ID: <eb1c75afc92@example.com>
+References: <681fb71bcf6@example.com>
+    <de34bbcfcf7@example.com>
+In-Reply-To: <de34bbcfcf7@example.com>
+To: Y <y@example.com>
+Cc: list@example.com
+MIME-Version: 1.0
+Content-Type: text/plain
+
+def
+EOM
+my %ibx = map {
+	$_ => create_inbox "v$_-event_step", version => $_,
+			indexlevel => 'medium', sub {
+		my ($im, $ibx) = @_;
+		$im->add($one) or xbail 'add 1';
+		$im->add($two) or xbail 'add 2';
+	};
+} (1, 2);
+write_file '>', my $cfgpath = "$tmpdir/config", <<EOM;
+[publicinbox "list"]
+	address = list\@example.com
+        url = http://example.com/list
+	inboxdir = $ibx{1}->{inboxdir}
+[publicinbox "list2"]
+	address = list2\@example.com
+        url = http://example.com/list2
+	inboxdir = $ibx{2}->{inboxdir}
+EOM
+my $env = { TMPDIR => $tmpdir, PI_CONFIG => $cfgpath };
+my $client = sub {
+	my ($cb) = @_;
+	my $res = $cb->(GET('/list/de34bbcfcf7@example.com/'));
+	is $res->code, 300, 'v1 300';
+	$res = $cb->(GET('/list2/de34bbcfcf7@example.com/'));
+	is $res->code, 300, 'v2 300';
+};
+
+test_httpd($env, $client);
+{
+	require PublicInbox::WWW;
+	my $cfg = PublicInbox::Config->new($cfgpath);
+	my $www = PublicInbox::WWW->new($cfg);
+	test_psgi(sub { $www->call(@_) }, $client);
+}
+done_testing;
next prev parent reply	other threads:[~2025-10-21  3:03 UTC|newest]
Thread overview: 8+ messages / expand[flat|nested]  mbox.gz  Atom feed  top
2025-10-13 21:41 httpd error: Can't locate object method "event_step" Kyle Meyer
2025-10-13 22:06 ` Eric Wong
2025-10-14  5:03   ` Kyle Meyer
2025-10-16  4:06     ` Kyle Meyer
2025-10-16  8:37       ` Eric Wong
2025-10-20 19:31         ` Eric Wong
2025-10-21  3:03           ` Eric Wong [this message]
2025-10-21  3:39             ` [PATCH] extmsg: fix Message-ID search for non-xap_helper users Kyle Meyer
find likely ancestor, descendant, or conflicting patches for this message:
( dfblob:282db1d2 dfblob:3e9b33a3 dfblob:941d11ae )
OR (
bs:"extmsg: fix Message-ID search for non-xap_helper users" )
(help)
Reply instructions:
You may reply publicly to this message via plain-text email
using any one of the following methods:
* Save the following mbox file, import it into your mail client,
and reply-to-all from there: mbox
Avoid top-posting and favor interleaved quoting:
https://en.wikipedia.org/wiki/Posting_style#Interleaved_style
List information: https://public-inbox.org/README
* Reply using the --to, --cc, and --in-reply-to
switches of git-send-email(1):
git send-email \
--in-reply-to=20251021030346.M346444@dcvr \
--to=e@80x24.org \
--cc=kyle@kyleam.com \
--cc=meta@public-inbox.org \
/path/to/YOUR_REPLY
https://kernel.org/pub/software/scm/git/docs/git-send-email.html
* If your mail client supports setting the In-Reply-To header
via mailto: links, try the mailto: link
Be sure your reply has a Subject: header at the top and a blank line
before the message body.
Code repositories for project(s) associated with this public inbox
https://80x24.org/public-inbox.git
This is a public inbox, see mirroring instructions
for how to clone and mirror all data and code used for this inbox;
as well as URLs for read-only IMAP folder(s) and NNTP newsgroup(s).
